File: src/api/ai-provenance.ts
/**
 * AI provenance API (design D22/D23, task 2.6).
 *
 * Recording is a SEPARATE invoke that runs after a domain create succeeded —
 * a provenance failure must never roll back (or block) a created card, and a
 * failed generation must never have anything to record.
 */

import { invokeCommand } from "../lib/tauri";

export interface AiProvenanceRecord {
  id: string;
  target_kind: string;
  target_id: string;
  task_id: string;
  provider: string;
  model?: string | null;
  model_class?: string | null;
  input_fingerprint?: string | null;
  created_at: string;
  metadata_json?: string | null;
}

export interface RecordAiProvenanceInput {
  /** What was created, e.g. "learning_item". */
  targetKind: string;
  /** Id of the created object. */
  targetId: string;
  /** Task that proposed it, e.g. "learn-this". */
  taskId: string;
  /** Provider id from the run result (`AITaskResult.providerId`). */
  provider: string;
  model?: string;
  modelClass?: string;
  /** fnv1a of the source passage (see `providers/types.ts` `fnv1aHash`). */
  inputFingerprint?: string;
  /** Selection context payload (passage, selectionContext, documentId, cardType). */
  metadata?: Record<string, unknown>;
}

export async function recordAiProvenance(
  input: RecordAiProvenanceInput
): Promise<AiProvenanceRecord> {
  return await invokeCommand<AiProvenanceRecord>("record_ai_provenance", {
    targetKind: input.targetKind,
    targetId: input.targetId,
    taskId: input.taskId,
    provider: input.provider,
    model: input.model,
    modelClass: input.modelClass,
    inputFingerprint: input.inputFingerprint,
    metadataJson: input.metadata,
  });
}

export async function getAiProvenance(
  targetKind: string,
  targetId: string
): Promise<AiProvenanceRecord[]> {
  return await invokeCommand<AiProvenanceRecord[]>("get_ai_provenance", {
    targetKind,
    targetId,
  });
}

File: src-tauri/src/commands/ai_provenance.rs
//! AI provenance commands (design D22/D23, task 2.6).
//!
//! Recording happens from the TS acceptance flow AFTER a domain create
//! succeeded (separate invoke, so a provenance failure can never roll back a
//! created card). `metadata_json` carries the original passage + selection
//! context payload for source navigation.

use crate::database::{AiProvenance, AiProvenanceRepository, Repository};
use crate::error::Result;
use tauri::State;

/// Record provenance for an accepted AI-generated object.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn record_ai_provenance(
    target_kind: String,
    target_id: String,
    task_id: String,
    provider: String,
    model: Option<String>,
    model_class: Option<String>,
    input_fingerprint: Option<String>,
    metadata_json: Option<serde_json::Value>,
    repo: State<'_, Repository>,
) -> Result<AiProvenance> {
    let repository = AiProvenanceRepository::new(repo.pool().clone());
    repository
        .insert_provenance(&AiProvenance {
            id: String::new(),
            target_kind,
            target_id,
            task_id,
            provider,
            model,
            model_class,
            input_fingerprint,
            created_at: String::new(),
            metadata_json: metadata_json.map(|value| value.to_string()),
        })
        .await
}

/// Get every provenance record attached to a target, oldest first.
#[tauri::command]
pub async fn get_ai_provenance(
    target_kind: String,
    target_id: String,
    repo: State<'_, Repository>,
) -> Result<Vec<AiProvenance>> {
    let repository = AiProvenanceRepository::new(repo.pool().clone());
    repository
        .get_provenance_for_target(&target_kind, &target_id)
        .await
}
